File: sknetwork/linkpred/nn.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created in March 2023
@author: Thomas Bonald <thomas.bonald@telecom-paris.fr>
"""
from typing import Optional, Union

import numpy as np
from scipy import sparse

from sknetwork.linkpred.base import BaseLinker
from sknetwork.embedding.base import BaseEmbedding
from sknetwork.linalg.normalizer import normalize
from sknetwork.utils.check import check_n_neighbors
from sknetwork.utils.format import get_adjacency


class NNLinker(BaseLinker):
    """Link prediction by nearest neighbors in the embedding space, using cosine similarity.

    For bipartite graphs, predict links between rows and columns only.

    Parameters
    ----------
    n_neighbors : int
        Number of nearest neighbors. If ``None``, all nodes are considered.
    threshold : float
        Threshold on cosine similarity. Only links above this threshold are kept.
    embedding_method : :class:`BaseEmbedding`
        Embedding method used to represent nodes in vector space.
        If ``None`` (default), use identity.

    Attributes
    ----------
    links_ : sparse.csr_matrix
        Link matrix.

    Example
    -------
    >>> from sknetwork.linkpred import NNLinker
    >>> from sknetwork.data import karate_club
    >>> linker = NNLinker(n_neighbors=5, threshold=0.5)
    >>> graph = karate_club(metadata=True)
    >>> adjacency = graph.adjacency
    >>> links = linker.fit_predict(adjacency)
    >>> links.shape
    (34, 34)
    """
    def __init__(self, n_neighbors: Optional[int] = 10, threshold: float = 0,
                 embedding_method: Optional[BaseEmbedding] = None):
        super(NNLinker, self).__init__()
        self.n_neighbors = n_neighbors
        self.threshold = threshold
        self.embedding_method = embedding_method
        self.bipartite = None

    def _fit_core(self, embedding, mask):
        n = embedding.shape[0]
        n_row = len(mask)
        if n_row < n:
            # bipartite graphs
            index_col = np.arange(n_row, n)
            n_col = n - n_row
        else:
            index_col = np.arange(n)
            n_col = n
        n_neighbors = check_n_neighbors(self.n_neighbors, len(index_col))

        row = []
        col = []
        data = []

        for i in np.flatnonzero(mask):
            vector = embedding[i]
            if sparse.issparse(vector):
                vector = vector.toarray().ravel()
            similarities = embedding[index_col].dot(vector)
            nn = np.argpartition(-similarities, n_neighbors)[:n_neighbors]
            mask_nn = np.zeros(n_col, dtype=bool)
            mask_nn[nn] = 1
            mask_nn[similarities < self.threshold] = 0
            nn = np.flatnonzero(mask_nn)

            row += len(nn) * [i]
            col += list(nn)
            data += list(similarities[nn])

        links = sparse.csr_matrix((data, (row, col)), shape=(n_row, n_col))

        return links

    def fit(self, input_matrix: Union[sparse.csr_matrix, np.ndarray], index: Optional[np.ndarray] = None) -> 'NNLinker':
        """Link prediction by nearest neighbors in the embedding space, using cosine similarity

        Parameters
        ----------
        input_matrix : sparse.csr_matrix, np.ndarray
            Adjacency matrix or biadjacency matrix of the graph.
        index : np.ndarray
            Index of source nodes to consider. If ``None``, the links are predicted for all nodes.

        Returns
        -------
        self: :class:`NN`
        """
        n_row, _ = input_matrix.shape

        adjacency, self.bipartite = get_adjacency(input_matrix)

        if index is None:
            index = np.arange(n_row)
        mask = np.zeros(n_row, dtype=bool)
        mask[index] = 1

        if self.embedding_method is None:
            embedding = adjacency
        else:
            embedding = self.embedding_method.fit_transform(adjacency)

        embedding = normalize(embedding, p=2)
        links = self._fit_core(embedding, mask)

        self.links_ = links

        return self

File: sknetwork-rs/src/linkpred/nn.rs
use sprs::{CsMat, TriMat};

use crate::embedding::spectral::Spectral;
use crate::utils::check::check_n_neighbors;
use crate::utils::format::{MatrixInput, get_adjacency};

/// Embedding backend used by [`NNLinker`].
#[derive(Debug, Clone)]
pub enum EmbeddingMethod {
    /// Spectral embedding via [`Spectral`].
    Spectral(Spectral),
}

#[derive(Debug, Clone)]
enum EmbeddingData {
    Dense(Vec<Vec<f64>>),
    Sparse(CsMat<f64>),
}

/// Errors raised by [`NNLinker`].
#[derive(Debug, Clone, PartialEq)]
pub enum NNError {
    /// Adjacency extraction or embedding fit failed.
    InvalidInput,
    /// `predict` or `fit_predict` called before `fit`.
    NotFitted,
}

/// Nearest-neighbor link-prediction estimator from node embeddings.
#[derive(Debug, Clone)]
pub struct NNLinker {
    /// Number of neighbors per row (`None` uses all columns).
    pub n_neighbors: Option<usize>,
    /// Minimum cosine similarity to keep a predicted link.
    pub threshold: f64,
    /// Optional embedding method (`None` uses normalized adjacency rows).
    pub embedding_method: Option<EmbeddingMethod>,
    /// Whether the last fit used a bipartite layout.
    pub bipartite: bool,
    /// Fitted link-score matrix.
    pub links: Option<CsMat<f64>>,
}

impl Default for NNLinker {
    fn default() -> Self {
        Self {
            n_neighbors: Some(10),
            threshold: 0.0,
            embedding_method: None,
            bipartite: false,
            links: None,
        }
    }
}

impl NNLinker {
    /// Creates a nearest-neighbor linker with explicit settings.
    ///
    /// # Arguments
    /// - `n_neighbors`: Neighbors per row (`None` uses all columns).
    /// - `threshold`: Minimum similarity to keep a predicted link.
    /// - `embedding_method`: Optional embedding backend.
    pub fn new(
        n_neighbors: Option<usize>,
        threshold: f64,
        embedding_method: Option<EmbeddingMethod>,
    ) -> Self {
        Self {
            n_neighbors,
            threshold,
            embedding_method,
            ..Self::default()
        }
    }

    fn normalize_rows_dense(x: &mut [Vec<f64>]) {
        for row in x {
            let norm = row.iter().map(|v| v * v).sum::<f64>().sqrt();
            if norm > 0.0 {
                for v in row {
                    *v /= norm;
                }
            }
        }
    }

    fn normalize_rows_sparse(x: &CsMat<f64>) -> CsMat<f64> {
        let (n, m) = x.shape();
        let mut tri = TriMat::<f64>::new((n, m));
        for (i, row) in x.outer_iterator().enumerate() {
            let norm = row.data().iter().map(|v| v * v).sum::<f64>().sqrt();
            if norm > 0.0 {
                for (j, v) in row.iter() {
                    tri.add_triplet(i, j, v / norm);
                }
            } else {
                for (j, v) in row.iter() {
                    tri.add_triplet(i, j, *v);
                }
            }
        }
        tri.to_csr::<usize>()
    }

    fn top_k_indices(sims: &[f64], k: usize) -> Vec<usize> {
        let kk = k.min(sims.len());
        if kk == 0 {
            return Vec::new();
        }
        let mut idx: Vec<usize> = (0..sims.len()).collect();
        if kk < idx.len() {
            idx.select_nth_unstable_by(kk - 1, |&a, &b| {
                sims[b]
                    .partial_cmp(&sims[a])
                    .unwrap_or(std::cmp::Ordering::Equal)
            });
            idx.truncate(kk);
        }
        idx
    }

    fn fit_core(&self, embedding: &EmbeddingData, mask: &[bool], n_row: usize) -> CsMat<f64> {
        let n = match embedding {
            EmbeddingData::Dense(x) => x.len(),
            EmbeddingData::Sparse(x) => x.rows(),
        };
        let (index_col, n_col) = if n_row < n {
            ((n_row..n).collect::<Vec<_>>(), n - n_row)
        } else {
            ((0..n).collect::<Vec<_>>(), n)
        };
        let mut global_to_local = vec![usize::MAX; n];
        for (pos, &j) in index_col.iter().enumerate() {
            global_to_local[j] = pos;
        }
        let requested = self.n_neighbors.unwrap_or(n_col);
        let n_neighbors = if n_row < n {
            // Bipartite row->col predictions do not require self-exclusion.
            requested.min(n_col)
        } else {
            check_n_neighbors(requested, n_col)
        };

        let mut tri = TriMat::<f64>::new((n_row, n_col));

        match embedding {
            EmbeddingData::Dense(x) => {
                for i in 0..n_row {
                    if !mask[i] {
                        continue;
                    }
                    let vector = &x[i];
                    let mut sims = vec![0.0; n_col];
                    for (pos, &j) in index_col.iter().enumerate() {
                        let mut s = 0.0;
                        for c in 0..vector.len() {
                            s += x[j][c] * vector[c];
                        }
                        sims[pos] = s;
                    }
                    let nn = Self::top_k_indices(&sims, n_neighbors);
                    for &jpos in &nn {
                        if sims[jpos] >= self.threshold {
                            tri.add_triplet(i, jpos, sims[jpos]);
                        }
                    }
                }
            }
            EmbeddingData::Sparse(x) => {
                let csc = x.to_csc();
                for i in 0..n_row {
                    if !mask[i] {
                        continue;
                    }
                    let mut sims = vec![0.0; n_col];
                    if let Some(row_i) = x.outer_view(i) {
                        // Accumulate similarities through shared nonzero features.
                        for (&feature, &val_i) in row_i.indices().iter().zip(row_i.data().iter()) {
                            if let Some(col_view) = csc.outer_view(feature) {
                                for (&j, &val_j) in col_view
                                    .indices()
                                    .iter()
                                    .zip(col_view.data().iter())
                                {
                                    let pos = global_to_local[j];
                                    if pos != usize::MAX {
                                        sims[pos] += val_i * val_j;
                                    }
                                }
                            }
                        }
                    }
                    let mut nn = Self::top_k_indices(&sims, n_neighbors);
                    // Keep Python-like behavior when threshold <= 0: allow filling top-k with zeros.
                    if self.threshold <= 0.0 && nn.len() < n_neighbors {
                        for pos in 0..n_col {
                            if nn.len() >= n_neighbors {
                                break;
                            }
                            if !nn.contains(&pos) && sims[pos] == 0.0 {
                                nn.push(pos);
                            }
                        }
                    }
                    for &jpos in &nn {
                        if sims[jpos] >= self.threshold {
                            tri.add_triplet(i, jpos, sims[jpos]);
                        }
                    }
                }
            }
        }
        tri.to_csr::<usize>()
    }

    /// Fits link scores from embeddings or normalized adjacency.
    ///
    /// # Arguments
    /// - `input_matrix`: Sparse adjacency or biadjacency input.
    /// - `index`: Optional row indices to score (`None` scores all rows).
    ///
    /// # Errors
    /// Returns [`NNError::InvalidInput`] when adjacency or embedding fit fails.
    pub fn fit(
        &mut self,
        input_matrix: &CsMat<f64>,
        index: Option<&[usize]>,
    ) -> Result<(), NNError> {
        let n_row = input_matrix.rows();
        let (adjacency, bip) = get_adjacency(
            MatrixInput::Sparse(input_matrix.to_owned()),
            true,
            false,
            false,
            false,
        )
        .map_err(|_| NNError::InvalidInput)?;
        self.bipartite = bip;
        let mut mask = vec![false; n_row];
        if let Some(ix) = index {
            for &i in ix {
                if i < n_row {
                    mask[i] = true;
                }
            }
        } else {
            for m in &mut mask {
                *m = true;
            }
        }

        let embedding = match &mut self.embedding_method {
            None => EmbeddingData::Sparse(Self::normalize_rows_sparse(&adjacency)),
            Some(EmbeddingMethod::Spectral(method)) => {
                let mut emb = method
                    .fit_transform(&adjacency, false)
                    .map_err(|_| NNError::InvalidInput)?;
                Self::normalize_rows_dense(&mut emb);
                EmbeddingData::Dense(emb)
            }
        };
        let links = self.fit_core(&embedding, &mask, n_row);
        self.links = Some(links);
        Ok(())
    }

    /// Fits link scores and returns the predicted link matrix.
    ///
    /// # Errors
    /// Returns:
    /// - [`NNError::InvalidInput`] when [`NNLinker::fit`] fails.
    /// - [`NNError::NotFitted`] if links were not stored after fit.
    pub fn fit_predict(
        &mut self,
        input_matrix: &CsMat<f64>,
        index: Option<&[usize]>,
    ) -> Result<CsMat<f64>, NNError> {
        self.fit(input_matrix, index)?;
        self.links.clone().ok_or(NNError::NotFitted)
    }

    /// Returns the fitted link-score matrix.
    ///
    /// # Errors
    /// Returns [`NNError::NotFitted`] when called before `fit`.
    pub fn predict(&self) -> Result<CsMat<f64>, NNError> {
        self.links.clone().ok_or(NNError::NotFitted)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::data::test_graphs::{test_digraph, test_graph};
    use crate::utils::neighbors::get_degrees;

    fn test_bigraph() -> CsMat<f64> {
        let mut tri = TriMat::<f64>::new((6, 8));
        for i in 0..6 {
            tri.add_triplet(i, i % 8, 1.0);
            tri.add_triplet(i, (i + 2) % 8, 1.0);
        }
        tri.to_csr::<usize>()
    }

    #[test]
    fn test_link_prediction() {
        for input_matrix in [test_graph(), test_digraph(), test_bigraph()] {
            let n_neighbors = 3usize;
            let threshold = 0.5;
            let mut algo = NNLinker::new(Some(n_neighbors), threshold, None);
            let links = algo.fit_predict(&input_matrix, None).unwrap();
            assert_eq!(links.shape(), input_matrix.shape());
            assert!(get_degrees(&links, false).iter().all(|&d| d <= n_neighbors));
            assert!(links.data().iter().all(|&x| x >= threshold));

            let spectral = Spectral::new(5, "rw", -1.0, true);
            let mut algo = NNLinker::new(
                Some(n_neighbors),
                0.0,
                Some(EmbeddingMethod::Spectral(spectral)),
            );
            let links = algo.fit_predict(&input_matrix, None).unwrap();
            assert_eq!(links.shape(), input_matrix.shape());
        }
    }

    #[test]
    fn test_predict_not_fitted() {
        let algo = NNLinker::default();
        assert!(matches!(algo.predict(), Err(NNError::NotFitted)));
    }
}
